Outline messaging scheme between WebProcess and UIProcess for sidePanel / sidebarAction.

AriYoung00 opened the pull request:
#### 1e8ea6e4777297ce82e6c911caa7cce2cc32e6a9
<pre>
Outline messaging scheme between WebProcess and UIProcess for sidePanel / sidebarAction.
<a href="https://webkit.org/b/277169">https://webkit.org/b/277169</a>
<a href="https://rdar.apple.com/132595403">rdar://132595403</a>

Reviewed by Timothy Hatcher.

This PR adds several messages and receiver functions on
WebExtensionContext, and one associated serializable type, in order to
facilitate the messaging between the Web and UI processes necessary for
sidePanel / sidebarAction.

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionSidebarParameters.serialization.in
* Source/WebKit/DerivedSources.make: Add
  WebExtensionWidebarParameters.serialization.in
* Source/WebKit/Shared/Extensions/WebExtensionSidebarParameters.h:
  Added.
* Source/WebKit/Shared/Extensions/WebExtensionSidebarParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm: Added.
(WebKit::WebExtensionContext::sidebarOpen): Added stub.
(WebKit::WebExtensionContext::sidebarClose): Added stub.
(WebKit::WebExtensionContext::sidebarIsOpen): Added stub.
(WebKit::WebExtensionContext::sidebarToggle): Added stub.
(WebKit::WebExtensionContext::sidebarSetIcon): Added stub.
(WebKit::WebExtensionContext::sidebarGetTitle): Added stub.
(WebKit::WebExtensionContext::sidebarSetTitle): Added stub.
(WebKit::WebExtensionContext::sidebarGetOptions): Added stub.
(WebKit::WebExtensionContext::sidebarSetOptions): Added stub.
(WebKit::WebExtensionContext::isSidebarMessageAllowed): Added stub which
returns false always.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Added
  declaration of sidebar message receivers.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
  Added sidebar messages.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added
  WebExtensionSidebarParameters.h and
WebExtensionSidebarParameters.serialization.in
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::WebExtensionAPISidebarAction::close): No change.
(WebKit::WebExtensionAPISidebarAction::toggle): No change.
(WebKit::WebExtensionAPISidebarAction::isOpen): Add `details` object
which was originally omitted erroneously.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h: Amend signature of `isOpen` to include `details` object.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl: Amend signature of `isOpen` to include `details` object.

Canonical link: <a href="https://commits.webkit.org/281519@main">https://commits.webkit.org/281519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be227265bc9a5fe000772ed0b76a8663710a0e42

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59914 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63832 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10442 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10664 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48558 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7281 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61944 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36628 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51902 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29430 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33334 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9363 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55262 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65564 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55904 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51881 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56057 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/13342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3180 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9016 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35075 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36157 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Shared/Extensions/WebExtensionSidebarParameters.h:
```diff
@@ -0,0 +1,40 @@
+namespace WebKit {
+
+struct WebExtensionSidebarParameters {
+    bool enabled;
```
`bool enabled { false }`;

AriYoung00 replied:
Shoot, I knew I was forgetting something from yesterday -- should be fixed now.

kiaraarose commented on Source/WebKit/DerivedSources.make:
```diff
@@ -611,6 +611,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	Shared/Extensions/WebExtensionMatchedRuleParameters.serialization.in \
 	Shared/Extensions/WebExtensionMenuItem.serialization.in \
 	Shared/Extensions/WebExtensionMessageSenderParameters.serialization.in \
+    Shared/Extensions/WebExtensionSidebarParameters.serialization.in \
```
indentation is off here

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -0,0 +1,89 @@
+
+}
+
+void WebExtensionContext::sidebarIsOpen(const std::optional<WebExtensionWindowIdentifier> windowIdentifier, CompletionHandler<void(Expected<bool, WebExtensionError> &&)>&& tabIdentifier)
```
`CompletionHandler<void(Expected<bool, WebExtensionError>&&)>&& completionHandler`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -0,0 +1,89 @@
+
+}
+
+void WebExtensionContext::sidebarToggle(CompletionHandler<void(Expected<void, WebExtensionError> &&)>&& completionHandler)
```
nit: `Expected<void, WebExtensionError>&&` i think convention is for the `&`'s to go with the type

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -48,6 +48,7 @@
 #include "WebExtensionMenuItem.h"
 #include "WebExtensionMessagePort.h"
 #include "WebExtensionPortChannelIdentifier.h"
+#include "WebExtensionSidebarParameters.h"
```
should this be guarded?

AriYoung00 replied:
I think it might not be necessary since the contents of the included file are guarded, but I'll go ahead and add one anyways.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
```diff
@@ -116,6 +116,19 @@ messages -> WebExtensionContext {
+    [EnabledIf='isSidebarMessageAllowed()'] SidebarGetTitle(std::optional<WebKit::WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebKit::WebExtensionTabIdentifier> tabIdentifier) -> (Expected<String, WebKit::WebExtensionError> result)
+    [EnabledIf='isSidebarMessageAllowed()'] SidebarSetTitle(std::optional<WebKit::WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebKit::WebExtensionTabIdentifier> tabIdentifier, std::optional<String> title) -> (Expected<void, WebKit::WebExtensionError> result)
+    [EnabledIf='isSidebarMessageAllowed()'] SidebarSetIcon(std::optional<WebKit::WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebKit::WebExtensionTabIdentifier> tabIdentifier, String iconDictionaryJSON) -> (Expected<void, WebKit::WebExtensionError> result)
+#endif
```
we tend to add a comment for the enable flag around a long block of code
`#endif // ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm:
```diff
@@ -0,0 +1,89 @@
+
+}
+
+void WebExtensionContext::sidebarSetIcon(const std::optional<WebExtensionWindowIdentifier> windowIdentifier, const std::optional<WebExtensionTabIdentifier> tabIdentifier, const String &iconDictionaryJSON, CompletionHandler<void(Expected<void, WebExtensionError> &&)>&& completionHandler)
```
same for these as well
`CompletionHandler<void(Expected<void, WebExtensionError>&&`

kiaraarose replied:
ditto `const String& iconDictionaryJSON`
but also would `iconJSON` be a better name?

AriYoung00 replied:
Probably so -- I'll change that as well

AriYoung00 committed:
Outline messaging scheme between WebProcess and UIProcess for sidePanel / sidebarAction.
https://webkit.org/b/277169
rdar://132595403

Reviewed by Timothy Hatcher.

This PR adds several messages and receiver functions on
WebExtensionContext, and one associated serializable type, in order to
facilitate the messaging between the Web and UI processes necessary for
sidePanel / sidebarAction.

* Source/WebKit/DerivedSources-input.xcfilelist: Add
  WebExtensionSidebarParameters.serialization.in
* Source/WebKit/DerivedSources.make: Add
  WebExtensionWidebarParameters.serialization.in
* Source/WebKit/Shared/Extensions/WebExtensionSidebarParameters.h:
  Added.
* Source/WebKit/Shared/Extensions/WebExtensionSidebarParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPISidebarCocoa.mm: Added.
(WebKit::WebExtensionContext::sidebarOpen): Added stub.
(WebKit::WebExtensionContext::sidebarClose): Added stub.
(WebKit::WebExtensionContext::sidebarIsOpen): Added stub.
(WebKit::WebExtensionContext::sidebarToggle): Added stub.
(WebKit::WebExtensionContext::sidebarSetIcon): Added stub.
(WebKit::WebExtensionContext::sidebarGetTitle): Added stub.
(WebKit::WebExtensionContext::sidebarSetTitle): Added stub.
(WebKit::WebExtensionContext::sidebarGetOptions): Added stub.
(WebKit::WebExtensionContext::sidebarSetOptions): Added stub.
(WebKit::WebExtensionContext::isSidebarMessageAllowed): Added stub which
returns false always.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Added
  declaration of sidebar message receivers.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
  Added sidebar messages.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added
  WebExtensionSidebarParameters.h and
WebExtensionSidebarParameters.serialization.in
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPISidebarActionCocoa.mm:
(WebKit::WebExtensionAPISidebarAction::close): No change.
(WebKit::WebExtensionAPISidebarAction::toggle): No change.
(WebKit::WebExtensionAPISidebarAction::isOpen): Add `details` object
which was originally omitted erroneously.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPISidebarAction.h: Amend signature of `isOpen` to include `details` object.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPISidebarAction.idl: Amend signature of `isOpen` to include `details` object.

Canonical link: https://commits.webkit.org/281519@main
